### DISCUSSION TITLE
Is it possible to use GridSearchCV for hierarchical data with Reconciler?

### DISCUSSION THREAD

**Discussion by anthonygiorgio97** [alan-turing-institute/sktime#4101] [Q&A]
Hi,

I have a hierarchical time series dataframe and for each time series I want to find the best forecaster from a list running a GridSearchCV using a params grid space. Then I want to reconcile the result with the `ReconcilerForecaster ` class. 

If I insert the reconciler forecaster into the forecaster parameter of `ForecastingGridSearchCV ` I have a message that this is not supported.

Another way, I think, is to fit the `ForecastingGridSearchCV ` for each series separately. But then how can I assign the best forecaster I found for each time series to the `ReconcilerForecaster ` ?

**Comment by fkiraly** [comment]
Strange, this should work - in theory (and my expectation), you should be able to combine the above-mentioned wrappers and any `sktime` forecaster arbitrarily.

Metrics, grid search, reconciler, individual forecasters, all should work out-of-the-box with hierarchical data.

There was a bug with `ReconcilerForecaster` which @ciaran-g fixed in the latest release 0.15.1.

In case something breaks for you, we would appreciate some short and self-contained code with dummy data that has the error you are experiencing, so we can debug.

FYI some people who come to my mind as having worked recently on hierarchical functionality: @ciaran-g, @danbartl, @KishManani.

**Comment by fkiraly** [comment]
Regarding your second question:

> Another way, I think, is to fit the `ForecastingGridSearchCV ` for each series separately. But then how can I assign the best forecaster I found for each time series to the `ReconcilerForecaster ` ?

If you want to do a grid search *for each series separately*, you need to wrap the grid search in `ForecastByLevel`, otherwise the grid search will compute the aggregate score and will try to find a single parameter setting that is best for all series, together (in mean aggregation if you use metrics with default values).

If you have the `ReconcilerForecaster` *inside* the grid search, the errors will be computed after reconciliation, if it is *outside*, the error to optimize will be *before* reconciling.

I'm not sure whether there is currently a way to both (a) fit grid search parameter by series and (b) tune by using metrics that are computed *after* reconciliation.

It will be possible though once @vyomkeshvyas finishes the `HierarchyEnsembleForecaster` here: https://github.com/sktime/sktime/pull/3905
Then, you would use the `HierarchyEnsembleForecaster` to construct an ensemble where parameters can be accessed by node, wrap it in the `ReconcilerForecaster` (or pipeline with the `Reconciler` transformer), then wrap in grid search.

**Comment by anthonygiorgio97** [comment]
Here there is an example of my use case:


```python
from sktime.utils._testing.hierarchical import _make_hierarchical
from sktime.forecasting.exp_smoothing import ExponentialSmoothing
from sktime.forecasting.trend import PolynomialTrendForecaster
from sktime.forecasting.model_selection import ForecastingGridSearchCV, ExpandingWindowSplitter
from sktime.transformations.hierarchical.aggregate import Aggregator
from sktime.forecasting.reconcile import ReconcilerForecaster
from sktime.forecasting.compose import TransformedTargetForecaster

y = _make_hierarchical()
agg = Aggregator()
y_agg = agg.fit_transform(y)

param_grid = [{"forecaster": [ExponentialSmoothing()],
               "forecaster__trend": ['add','mul']
              },
              {"forecaster": [PolynomialTrendForecaster()],
               "forecaster__degree": [1,2]}
              ]

pipe = TransformedTargetForecaster(steps=[
    ("forecaster", ExponentialSmoothing())])

N_cv_fold = 2
step_cv = 1
fh = [1,2]

initial_window_cv_len = len(y_agg.index.get_level_values(2).unique()) - (N_cv_fold - 1) * step_cv - fh[-1]

cv = ExpandingWindowSplitter(
            initial_window = initial_window_cv_len,
            step_length = step_cv,
            fh = fh)

reconciler = ReconcilerForecaster(pipe, method="ols")

gscv = ForecastingGridSearchCV(
        forecaster=reconciler,
        param_grid=param_grid,
        cv=cv,
        n_jobs=-1,
        verbose = 1
        )

gscv.fit(y_agg)
```

However in this case, as you said, the grid search will compute the aggregate score and will try to find a single parameter setting that is best for all series, together and not the best forecaster for each series in the hierarchy (which is my goal)

Another approach would be to use a for cicle for the most granular hierarchy index and store best forecaster for each series in a dictionary (however it will not use the benefits of vectorization) as the following code. It would then be ideal if we could configure the forecasters for each series, for example in the dataframe of the image below extract from a tutorial on sktime site:

![image](https://user-images.githubusercontent.com/67336832/212328812-1399c426-a4ed-47a3-bfd0-d7ed38ff7ebf.png)

```python
param_grid = [{"forecaster": [ExponentialSmoothing()],
               "forecaster__trend": ['add','mul']
              },
              {"forecaster": [PolynomialTrendForecaster()],
               "forecaster__degree": [1,2]}
              ]

N_cv_fold = 2
step_cv = 1
fh = [1,2]

initial_window_cv_len = len(y_agg.index.get_level_values(2).unique()) - (N_cv_fold - 1) * step_cv - fh[-1]

hierarchy = list(set(list(zip(y_agg.index.get_level_values(0),
                              y_agg.index.get_level_values(1)))))

forecaster_dict = {}

for ts in hierarchy:

    y_ts = y_agg[(y_agg.index.get_level_values(0) == ts[0]) &
                 (y_agg.index.get_level_values(1) == ts[1])]

    pipe = TransformedTargetForecaster(steps=[
                    ("forecaster", ExponentialSmoothing())])

    cv = ExpandingWindowSplitter(
            initial_window = initial_window_cv_len,
            step_length = step_cv,
            fh = fh)
    
    gscv = ForecastingGridSearchCV(
                forecaster=pipe,
                param_grid=param_grid,
                cv=cv,
                n_jobs=-1,
                verbose = 1
                )

    gscv.fit(y_ts)
    forecaster_dict[ts] = gscv.best_forecaster_
```

**Reply by anthonygiorgio97** [reply]
Another question: here I use `ForecastByLevel`, as suggested, and compute reconciliation after. How can I access the best forecaster fitted for each time series at each hierarchy? If I use `gscv_bylevel.get_fitted_params()["forecasters"]` I obtain a pandas dataframe where forecaster is the entire pipeline and not the best_forecaster

```python
from sktime.utils._testing.hierarchical import _make_hierarchical
from sktime.forecasting.exp_smoothing import ExponentialSmoothing
from sktime.forecasting.trend import PolynomialTrendForecaster
from sktime.forecasting.model_selection import ForecastingGridSearchCV, ExpandingWindowSplitter
from sktime.transformations.hierarchical.aggregate import Aggregator
from sktime.forecasting.reconcile import ReconcilerForecaster
from sktime.forecasting.compose import TransformedTargetForecaster
from sktime.forecasting.compose import ForecastByLevel

y = _make_hierarchical()
agg = Aggregator()
y_agg = agg.fit_transform(y)

param_grid = [{"forecaster": [ExponentialSmoothing()],
               "forecaster__trend": ['add','mul']
              },
              {"forecaster": [PolynomialTrendForecaster()],
               "forecaster__degree": [1,2]}
              ]

pipe = TransformedTargetForecaster(steps=[
    ("forecaster", ExponentialSmoothing())])

N_cv_fold = 2
step_cv = 1
fh = [1,2]

initial_window_cv_len = len(y_agg.index.get_level_values(2).unique()) - (N_cv_fold - 1) * step_cv - fh[-1]

cv = ExpandingWindowSplitter(
            initial_window = initial_window_cv_len,
            step_length = step_cv,
            fh = fh)

gscv = ForecastingGridSearchCV(
        forecaster=pipe,
        param_grid=param_grid,
        cv=cv,
        n_jobs=-1,
        verbose = 1
        )

gscv_bylevel = ForecastByLevel(gscv, 'local')

gscv_bylevel.fit(y_agg)

reconciler = ReconcilerForecaster(gscv_bylevel, method="ols")

gscv_bylevel.get_fitted_params()["forecasters"]
```

![image](https://user-images.githubusercontent.com/67336832/212338196-d6a3f031-576d-44ab-ad77-20677223aced.png)

**Reply by fkiraly** [reply]
> If I use `gscv_bylevel.get_fitted_params()["forecasters"]` I obtain a pandas dataframe where forecaster is the entire pipeline and not the best_forecaster

You just need to continue unwrapping, either by looking at keys with multiple double underscores, or unwrapping the data frame. 

So, for instance, `gscv_bylevel.get_fitted_params()["forecasters"].loc[('__total', '__total'),'forecasters']` gives you the forecaster at the all total index, and you can further unwrap it with `gscv_bylevel.get_fitted_params()["forecasters"].loc[('__total', '__total'),'forecasters'].get_fitted_params()["forecaster__best_forecaster__forecaster"]`

Using keys only, for instance, `gscv_bylevel.get_fitted_params()["forecasters.loc[('__total', '__total'),forecasters]__forecaster__best_forecaster__forecaster"]`,
gives you the selected forecaster for the overall total.

I'm just noticing that there may be a bug with string based indexing, since only one index appears in the string keys, I need to look at this. Data frame based indexing should work though.

**Reply by fkiraly** [reply]
Update: there was a bug with `best_forecaster` etc not appearing in `get_fitted_params`, see here: https://github.com/sktime/sktime/pull/4102

Should be fixed in the next release. We had the last one yesterday, so it might be a couple weeks.

**Reply by anthonygiorgio97** [reply]
Thanks.

Then when I call the .predict() method, in this case the best forecaster is used for each time series, right? (in my example as follow:)

```python
best_forecasters = {}
for ts in gscv_bylevel.get_fitted_params()["forecasters"].index:
    best_forecasters[ts] = gscv_bylevel.get_fitted_params()["forecasters"] \
                                       .loc[(ts[0], ts[1]),'forecasters'] \
                                       .get_fitted_params()["forecaster__best_forecaster__forecaster"]
best_forecasters
```

```
{('__total', '__total'): ExponentialSmoothing(trend='add'),
 ('h0_0', '__total'): PolynomialTrendForecaster(),
 ('h0_0', 'h1_0'): ExponentialSmoothing(trend='mul'),
 ('h0_0', 'h1_1'): ExponentialSmoothing(trend='mul'),
 ('h0_0', 'h1_2'): ExponentialSmoothing(trend='add'),
 ('h0_0', 'h1_3'): PolynomialTrendForecaster(),
 ('h0_1', '__total'): PolynomialTrendForecaster(),
 ('h0_1', 'h1_0'): ExponentialSmoothing(trend='mul'),
 ('h0_1', 'h1_1'): PolynomialTrendForecaster(),
 ('h0_1', 'h1_2'): ExponentialSmoothing(trend='add'),
 ('h0_1', 'h1_3'): ExponentialSmoothing(trend='add')}
```

And what about the reconciliation applied on `gscv_bylevel`? Will it use the best forecasters for each series or it will re-do the gridsearch? (I ask this because when I run the following code I obtain the following verbose output. However it is strange because the "Fitting x folds for each of y candidates, totalling x*y fits" verbose output appear only 4 times, while there are 11 time series)

```python
reconciler = ReconcilerForecaster(gscv_bylevel, method="ols")
reconciler.fit(y_agg)
reconciler.predict([1,2])
```

```
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
```

Finally, from your experience, it is a better practice to apply reconciliation after the finding of the best forecaster for single series (as my example), or before wrapping it into the gridsearch (and fitting the reconciler constructed on `gscv_bylevel` directly instead of first fit the `gscv_bylevel`) ?

**Reply by fkiraly** [reply]
> Then when I call the .predict() method, in this case the best forecaster is used for each time series, right

Yes (per intention, if not it's a bug). A small caveat is on use of `update`, this may differ depending on the `update_behaviour` parameter, which determines whether the tuned parameters are also re-calculated on the larger data set, or not.

> And what about the reconciliation applied on `gscv_bylevel`? Will it use the best forecasters for each series or it will re-do the gridsearch? (I ask this because when I run the following code I obtain the following verbose output. However it is strange because the "Fitting x folds for each of y candidates, totalling x*y fits" verbose output appear only 4 times, while there are 11 time series)

It should first do the grid search for each of the 11 series, then make forecasts, then apply the reconciliation (since you selected the `"ols"` strategy, it should be the same as a pipeline with the `Reconciler`, so should result in precisely 11 grid searches).

It is odd that it appears only 4 times, and I wouldn't be able to say why the number 4 is special here. Are you sure you just get 4 printouts? If yes, it might be worth investigating and/or reporting as a bug.

> Finally, from your experience, it is a better practice to apply reconciliation after the finding of the best forecaster for single series (as my example), or before wrapping it into the gridsearch

The more parsimonious (simpler) model is the latter, i.e., just one parameter for all series, because it has less parameters. If you do the former (grid search per series), you should always compare it with the latter imo, since it the latter "the simpler version", therefore a partial baseline. Another term for it would be the "natural ablation". cf "ablation study".

Which one is better will depend on your data and of your choice of models. As a rule of thumb, you need more data to increase the "moral chance" that a model with more parameters will be better. The "data needed in terms of parameters" asymptotics may vary depending on model class, but typically you would expect you need (substantially) more effective sample size than model parameters for a good fit that generalizes on unseen data.

**Reply by anthonygiorgio97** [reply]
> It should first do the grid search for each of the 11 series, then make forecasts, then apply the reconciliation (since you selected the "ols" strategy, it should be the same as a pipeline with the Reconciler, so should result in precisely 11 grid searches).

Ok thanks. What if I don't want to redo the grid search and only apply the reconciliation on the best forecasters results for each series? It is possible?

> It is odd that it appears only 4 times, and I wouldn't be able to say why the number 4 is special here. Are you sure you just get 4 printouts? If yes, it might be worth investigating and/or reporting as a bug.

This is what I got. If you can confirm the same by running the code, maybe it should be investigated

**Reply by fkiraly** [reply]
>Ok thanks. What if I don't want to redo the grid search and only apply the reconciliation on the best forecasters results for each series? It is possible?

To clarify: I don't think it should be redoing any grid searches at all in `predict` if you choose `"ols"`. I.e., it should first do 11 grid searches to find the best forecaster per series, then forecast per series, then reconcile the forecasts.

> This is what I got. If you can confirm the same by running the code, maybe it should be investigated

Which of the above blocks does that? Could you maybe just re-post minimal code with the issue?

**Reply by anthonygiorgio97** [reply]
> It is odd that it appears only 4 times, and I wouldn't be able to say why the number 4 is special here. Are you sure you just get 4 printouts? If yes, it might be worth investigating and/or reporting as a bug.

I checked again. By restarting the kernel I do not have this issue again.

---

So this would be my final solution for my problem:

```python
from sktime.utils._testing.hierarchical import _make_hierarchical
from sktime.forecasting.exp_smoothing import ExponentialSmoothing
from sktime.forecasting.trend import PolynomialTrendForecaster
from sktime.forecasting.model_selection import ForecastingGridSearchCV, ExpandingWindowSplitter
from sktime.transformations.hierarchical.aggregate import Aggregator
from sktime.forecasting.reconcile import ReconcilerForecaster
from sktime.forecasting.compose import TransformedTargetForecaster
from sktime.forecasting.compose import ForecastByLevel

y = _make_hierarchical()
agg = Aggregator()
y_agg = agg.fit_transform(y) 

param_grid = [{"forecaster": [ExponentialSmoothing()],
               "forecaster__trend": ['add','mul']
              },
              {"forecaster": [PolynomialTrendForecaster()],
               "forecaster__degree": [1,2]}
              ]

pipe = TransformedTargetForecaster(steps=[
    ("forecaster", ExponentialSmoothing())])

N_cv_fold = 2
step_cv = 1
fh = [1,2]

initial_window_cv_len = len(y_agg.index.get_level_values(2).unique()) - (N_cv_fold - 1) * step_cv - fh[-1]

cv = ExpandingWindowSplitter(
            initial_window = initial_window_cv_len,
            step_length = step_cv,
            fh = fh)

gscv = ForecastingGridSearchCV(
        forecaster=pipe,
        param_grid=param_grid,
        cv=cv,
        n_jobs=-1,
        verbose = 1
        )

gscv_bylevel = ForecastByLevel(gscv, 'local')

reconciler = ReconcilerForecaster(gscv_bylevel, method="ols")

reconciler.fit(y_agg)
```
```
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
Fitting 2 folds for each of 4 candidates, totalling 8 fits
```

---

To access the best fitted forecaster for each time series hierarchy I can run the following code:

```python
best_forecasters = {}
for ts in reconciler.get_fitted_params()['forecaster__ForecastByLevel__forecasters'].index:
    best_forecasters[ts] = reconciler.get_fitted_params()['forecaster__ForecastByLevel__forecasters'] \
                                       .loc[(ts[0], ts[1]),'forecasters'] \
                                       .get_fitted_params()["forecaster__best_forecaster__forecaster"]
best_forecasters
```
```

{('__total', '__total'): ExponentialSmoothing(trend='add'),
 ('h0_0', '__total'): PolynomialTrendForecaster(),
 ('h0_0', 'h1_0'): ExponentialSmoothing(trend='mul'),
 ('h0_0', 'h1_1'): ExponentialSmoothing(trend='mul'),
 ('h0_0', 'h1_2'): ExponentialSmoothing(trend='add'),
 ('h0_0', 'h1_3'): PolynomialTrendForecaster(),
 ('h0_1', '__total'): PolynomialTrendForecaster(),
 ('h0_1', 'h1_0'): ExponentialSmoothing(trend='mul'),
 ('h0_1', 'h1_1'): PolynomialTrendForecaster(),
 ('h0_1', 'h1_2'): ExponentialSmoothing(trend='add'),
 ('h0_1', 'h1_3'): ExponentialSmoothing(trend='add')}
```

**Reply by jpss95** [reply]
This solution seemed to be suitable for my problem, where I wanted to take a local forecasting approach by finding the best model for each time series. When I tried it on a small scale, it worked perfectly, but when attempting to use it on a problem with many time series, it didn't scale.

I'm using a machine with multiple cores, and I'm providing the argument n_jobs=-1 to ForecastingGridSearchCV to make use of them. However, in reality, it doesn't utilize them, and all the work is done sequentially.

![image](https://github.com/sktime/sktime/assets/64998671/f762782a-7eaa-47bc-99c2-bf1cb8ee5c0e)

cc: @fkiraly

**Reply by fkiraly** [reply]
> However, in reality, it doesn't utilize them, and all the work is done sequentially.

`ForecastingGridSearchCV` parallelizes parameters, not time series instances in the hierarchy - we could write parallelization for the multiple series though.

**Reply by fkiraly** [reply]
or is this a more general issue with the tuners? Could you perhaps try the same without hierarchical data, and check your cores?

**Reply by jpss95** [reply]
Thanks for the clarification, Franz. It makes sense that ForecastingGridSearchCV focuses on hyperparameter optimization. But I believe that introducing a parallelization feature at the time series level could greatly enhance the efficiency of the library for users like me who deal with multiple time series.

**Comment by fkiraly** [comment]
Hm, what you did here is very interesting:

```python
best_forecasters = {}
for ts in gscv_bylevel.get_fitted_params()["forecasters"].index:
    best_forecasters[ts] = gscv_bylevel.get_fitted_params()["forecasters"] \
                                       .loc[(ts[0], ts[1]),'forecasters'] \
                                       .get_fitted_params()["forecaster__best_forecaster__forecaster"]
best_forecasters
```

that is, the data frame is filled with nested parameters if called with the name of the nested parameter.

Perhaps this should just happen by default if you call `gscv_bylevel.get_fitted_params()["best_forecaster"]` or similar, don't you think?

**Reply by anthonygiorgio97** [reply]
If I try this I get an error. However it will be very helpful if implemented

**Reply by fkiraly** [reply]
yes, I know. That's what I'm saying, it *should* just produce the data frame you made! Currently it doesn't.

**Reply by fkiraly** [reply]
i.e., I got the idea that it would be very helpful just to happen by default, in general, from looking at what you did. Not claiming that it currently behaves that way.

**Reply by fkiraly** [reply]
here is a prototype, @anthonygiorgio97:
https://github.com/sktime/sktime/pull/4107

would appreciate some feedback what you think!

**Comment by fkiraly** [comment]
> It would then be ideal if we could configure the forecasters for each series, for example in the dataframe of the image below extract from a tutorial on sktime site:

Can you explain what you mean here by "it would be ideal"?

* does the `ForecastByLevel` approach work for you? Or is that not what you want?
* or, do you think it's too "clunky", and you would prefer a parameter, say, in `ForecastingGridSearchCV`?

**Reply by anthonygiorgio97** [reply]
I'm sorry, but I was not aware of the existence of the `ForecastByLevel` approach. It is ok to use it for this goal!

**Reply by fkiraly** [reply]
oh, I mentioned it here in my first reply, it might have gotten lost in all the threads: https://github.com/sktime/sktime/discussions/4101#discussioncomment-4676681

**Reply by fkiraly** [reply]
confused - you also used it here in your own code?? https://github.com/sktime/sktime/discussions/4101#discussioncomment-4678153

**Reply by anthonygiorgio97** [reply]
Yes. Sorry, my mistake: I have posted my first reply and after I documented and tried the `ForecastByLevel` approach

**Reply by fkiraly** [reply]
ah, I see, so "was not aware" refers to your state of knowledge 4 hours ago. Makes sense. Would appreciate feedback in case you decide to try it out.

**Reply by fkiraly** [reply]
might also be good if we add some of the newer stuff to the hierarchical forecasting tutorial.

**Comment by daniel-torresc** [comment]
Hello,

I have tried to reproduce the code on this thread (see below) and had an issue when trying to access the `reconciler.forecaster.get_fitted_params()` method of the reconciler when fitted. The output is the following:

```
NotFittedError                            Traceback (most recent call last)
Cell In[17], line 1
----> 1 reconciler.forecaster.get_fitted_params()

File ~\AppData\Roaming\Python\Python39\site-packages\sktime\forecasting\base\_base.py:1256, in BaseForecaster.get_fitted_params(self, deep)
1254 # if self is not vectorized, run the default get_fitted_params
1255 if not getattr(self, "_is_vectorized", False):
-> 1256     return super(BaseForecaster, self).get_fitted_params(deep=deep)
1258 # otherwise, we delegate to the instances' get_fitted_params
1259 # instances' parameters are returned at dataframe-slice-like keys
1260 fitted_params = {}

File ~\AppData\Roaming\Python\Python39\site-packages\sktime\base\_base.py:417, in BaseEstimator.get_fitted_params(self, deep)
    386 """Get fitted parameters.
    387 
    388 State required:
(...)
    414       e.g., `[componentname]__[componentcomponentname]__[paramname]`, etc
    415 """
    416 if not self.is_fitted:
--> 417     raise NotFittedError(
    418         f"estimator of type {type(self).__name__} has not been "
    419         "fitted yet, please call fit on data before get_fitted_params"
    420     )
    422 # collect non-nested fitted params of self
    423 fitted_params = self._get_fitted_params()

NotFittedError: estimator of type ForecastByLevel has not been fitted yet, please call fit on data before get_fitted_params
```

I have come up with a "partial solution" which is fitting again the forecaster, in addition to the previous reconciler fit. This is:

**Step 1**: Calling `reconciler.fit(y_train_agg)` (here when trying to access `reconciler.forecaster.get_fitted_params()` it throws the exception)
**Step 2**: Additionally, execute `reconciler.forecaster.fit(y_train_agg)` (ok when accessing `get_fitted_params()` after executing this fit)

I have tried this with sktime 0.17.1, 0.17.2 and the latest 0.18 versions and all had the same issue.

Is there any way to access `get_fitted_params()` without needing to explicitly fit the forecaster (step 2) again? I would like to access it once the reconciler is fitted.

Has anyone been able to execute this code without having errors? I understand the solution to my question is what is proposed in the thread, but I am having the issue listed above.

The code I have tried is the one below:

```python
from sktime.utils._testing.hierarchical import _make_hierarchical
from sktime.forecasting.exp_smoothing import ExponentialSmoothing
from sktime.forecasting.trend import PolynomialTrendForecaster
from sktime.forecasting.model_selection import ForecastingGridSearchCV, ExpandingWindowSplitter
from sktime.transformations.hierarchical.aggregate import Aggregator
from sktime.forecasting.reconcile import ReconcilerForecaster
from sktime.forecasting.compose import TransformedTargetForecaster
from sktime.forecasting.compose import ForecastByLevel

y = _make_hierarchical()
agg = Aggregator()
y_agg = agg.fit_transform(y) 

param_grid = [{"forecaster": [ExponentialSmoothing()],
               "forecaster__trend": ['add','mul']
              },
              {"forecaster": [PolynomialTrendForecaster()],
               "forecaster__degree": [1,2]}
              ]

pipe = TransformedTargetForecaster(steps=[
    ("forecaster", ExponentialSmoothing())])

N_cv_fold = 2
step_cv = 1
fh = [1,2]

initial_window_cv_len = len(y_agg.index.get_level_values(2).unique()) - (N_cv_fold - 1) * step_cv - fh[-1]

cv = ExpandingWindowSplitter(
            initial_window = initial_window_cv_len,
            step_length = step_cv,
            fh = fh)

gscv = ForecastingGridSearchCV(
        forecaster=pipe,
        param_grid=param_grid,
        cv=cv,
        n_jobs=-1,
        verbose = 1
        )

gscv_bylevel = ForecastByLevel(gscv, 'local')

reconciler = ReconcilerForecaster(gscv_bylevel, method="ols")

reconciler.fit(y_agg)

# reconciler.forecaster.fit(y_agg)  # If this line is uncommented, the error disappears

reconciler.forecaster.get_fitted_params()  # here we get the error when accessing ".get_fitted_params()"
```

Thank you.

**Reply by fkiraly** [reply]
@daniel-torresc, could you be kindly specific as to which piece of code (e.g., link or paste) give you which error?
If this is unexpected, would appreciate a bug report.
(I think we added the above case as a test case and it runs in CI)

**Reply by fkiraly** [reply]
apologies, just saw you pasted it. Let me try.

**Reply by fkiraly** [reply]
ok, here's the thing: that `reconciler.forecaster.get_fitted_params` fails is *not a bug*, because the `forecaster` is just a template, a clone is fitted.

However, if you call `reconciler.get_fitted_params`, it fails - that is a bug!

**Reply by fkiraly** [reply]
reported here: https://github.com/sktime/sktime/issues/4574

**Reply by daniel-torresc** [reply]
Thank you so much for the quick response! I will keep an eye on it and will check again when it's fixed.

**Reply by fkiraly** [reply]
fixed in 0.18.1, @daniel-torresc
